orion/primitives/tadgan.py

```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
+            'target_shape': self.target_shape,
+            'input_dim': input_dim,
+            'output_dim': output_dim,
+            'generator_reshape_dim': length // 2,
+            'generator_reshape_shape': (length // 2, output_dim),
+            'encoder_reshape_shape': self.latent_shape
+        })
         return kwargs
 
-    def _set_shapes(self):
+    def _set_shapes(self) -> None:
         self.encoder_input_shape = self.shape
         self.generator_input_shape = self.latent_shape
         self.critic_x_input_shape = self.target_shape
         self.critic_z_input_shape = self.latent_shape
 
-    def _build_tadgan(self, **kwargs):
+    def _build_tadgan(self, **kwargs) -> None:
         hyperparameters = self.hyperparameters.copy()
         hyperparameters.update(kwargs)
 
-        self.encoder = self._build_model(hyperparameters, self.layers_encoder,
-                                         self.encoder_input_shape)
-        self.generator = self._build_model(hyperparameters, self.layers_generator,
-                                           self.generator_input_shape)
-        self.critic_x = self._build_model(hyperparameters, self.layers_critic_x,
-                                          self.critic_x_input_shape)
-        self.critic_z = self._build_model(hyperparameters, self.layers_critic_z,
-                                          self.critic_z_input_shape)
-
-        self.generator.trainable = False
-        self.encoder.trainable = False
-
-        x = Input(shape=self.shape)
-        y = Input(shape=self.target_shape)
-        z = Input(shape=(self.latent_dim, 1))
+        # Models
+        self.encoder = self._build_model(
+            hyperparameters, self.layers_encoder, self.encoder_input_shape, name='encoder')
+        self.generator = self._build_model(
+            hyperparameters, self.layers_generator, self.generator_input_shape, name='generator')
+        self.critic_x = self._build_model(
+            hyperparameters, self.layers_critic_x, self.critic_x_input_shape, name='critic_x')
+        self.critic_z = self._build_model(
+            hyperparameters, self.layers_critic_z, self.critic_z_input_shape, name='critic_z')
+
+        # Optimizers
+        self.critic_x_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.critic_z_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.encoder_generator_optimizer = import_object(self.optimizer)(self.learning_rate)
+
+    def _format_losses(self, losses: list) -> dict:
+        """Format losses into dictionary mapping metric names to their current value."""
+        output = dict()
+        if self.detailed_losses:
+            for i in range(len(losses)):
+                for j in range(len(losses[i])):
+                    output[LOSS_NAMES[i][j]] = losses[i][j]
+        else:
+            for i in range(len(losses)):
+                output[LOSS_NAMES[i][0]] = losses[i][0]
+        return output
```

Comment:
minor: maybe add a space between the two lines for easy readability.

typically when you go back in indentation, we add a space.

Reply:
Added space!